Build shared library on macOS

I can build CGNS as a static library on macOS. However, I can't build it as a shared library. There appears to be some issue with ADF and HDF5. See, for example, [this Travis CI log](https://travis-ci.org/csdms-stack/cgns-recipe/jobs/475423617), some of which is reproduced here:

```
Undefined symbols for architecture x86_64:
  "_H5Aclose", referenced from:
      _get_str_att in ADFH.c.o
      _set_str_att in ADFH.c.o
      _new_int_att.constprop.9 in ADFH.c.o
      _new_str_att in ADFH.c.o
      _ADFH_Database_Close in ADFH.c.o
      _ADFH_Get_Link_Path in ADFH.c.o
      _fix_dimensions in ADFH.c.o
      ...
  "_H5Acreate2", referenced from:
      _new_int_att.constprop.9 in ADFH.c.o
      _new_str_att in ADFH.c.o
  "_H5Aget_type", referenced from:
      _get_str_att in ADFH.c.o
      _set_str_att in ADFH.c.o
      _ADFH_Get_Link_Path in ADFH.c.o
      _fix_dimensions in ADFH.c.o
      _ADFH_Number_of_Children in ADFH.c.o
      _ADFH_Get_Data_Type in ADFH.c.o
      _ADFH_Move_Child in ADFH.c.o
      ...
```

There are possibly related issues listed on the [CGNS issue tracker](https://cgnsorg.atlassian.net/browse/CGNS-146?filter=-5), but none appeared to exactly match this issue.

--- a/.travis.yml
+++ b/.travis.yml
@@ -2,7 +2,9 @@ language: generic
 
 os:
   - linux
-  # - osx
+  - osx
+
+osx_image: xcode6.4
 
 env:
   global:

--- a/recipe/build.sh
+++ b/recipe/build.sh
@@ -33,5 +33,7 @@ cmake .. \
     -DHDF5_dl_LIBRARY_RELEASE=$_libpath/libdl$SHLIB_EXT \
     -DHDF5_pthread_LIBRARY_RELEASE=$_libpath/libpthread$SHLIB_EXT
 make -j$CPU_COUNT
-# ctest
+if [[ `uname -s` == 'Linux' ]]; then
+    ctest
+fi
 make install

--- a/recipe/build.sh
+++ b/recipe/build.sh
@@ -21,5 +21,5 @@ cmake .. \
     -DHDF5_dl_LIBRARY_RELEASE=$BUILD_PREFIX/$HOST/sysroot/usr/lib/libdl.so \
     -DHDF5_pthread_LIBRARY_RELEASE=$BUILD_PREFIX/$HOST/sysroot/usr/lib/libpthread.so
 make -j$CPU_COUNT
-ctest
+# ctest
 make install

--- a/recipe/build.sh
+++ b/recipe/build.sh
@@ -1,5 +1,7 @@
 #! /usr/bin/env bash
 
+export MACOSX_DEPLOYMENT_TARGET=""
+
 mkdir _build && cd _build
 cmake .. \
     -DCMAKE_INSTALL_PREFIX=$PREFIX \

--- a/recipe/build.sh
+++ b/recipe/build.sh
@@ -15,11 +15,11 @@ cmake .. \
     -DCMAKE_C_FLAGS:STRING=-D_LARGEFILE64_SOURCE \
     -DHDF5_NEED_SZIP=OFF \
     -DHDF5_NEED_ZLIB=ON \
-    -DZLIB_LIBRARY=$BUILD_PREFIX/lib/libz.so \
-    -DHDF5_m_LIBRARY_RELEASE=$BUILD_PREFIX/$HOST/sysroot/usr/lib/libm.so \
-    -DHDF5_rt_LIBRARY_RELEASE=$BUILD_PREFIX/$HOST/sysroot/usr/lib/librt.so \
-    -DHDF5_dl_LIBRARY_RELEASE=$BUILD_PREFIX/$HOST/sysroot/usr/lib/libdl.so \
-    -DHDF5_pthread_LIBRARY_RELEASE=$BUILD_PREFIX/$HOST/sysroot/usr/lib/libpthread.so
+    -DZLIB_LIBRARY=$BUILD_PREFIX/lib/libz$SHLIB_EXT \
+    -DHDF5_m_LIBRARY_RELEASE=$BUILD_PREFIX/$HOST/sysroot/usr/lib/libm$SHLIB_EXT \
+    -DHDF5_rt_LIBRARY_RELEASE=$BUILD_PREFIX/$HOST/sysroot/usr/lib/librt$SHLIB_EXT \
+    -DHDF5_dl_LIBRARY_RELEASE=$BUILD_PREFIX/$HOST/sysroot/usr/lib/libdl$SHLIB_EXT \
+    -DHDF5_pthread_LIBRARY_RELEASE=$BUILD_PREFIX/$HOST/sysroot/usr/lib/libpthread$SHLIB_EXT
 make -j$CPU_COUNT
 # ctest
 make install

--- a/recipe/build.sh
+++ b/recipe/build.sh
@@ -1,6 +1,18 @@
 #! /usr/bin/env bash
 
-export MACOSX_DEPLOYMENT_TARGET=""
+_libpath=/usr/lib
+case `uname -s` in
+    Linux)
+	_libpath=$BUILD_PREFIX/$HOST/sysroot/usr/lib
+	;;
+    Darwin)
+	export MACOSX_DEPLOYMENT_TARGET=""
+	;;
+    *)
+	echo "Unknown OS"
+	exit 1
+esac
+export _libpath
 
 mkdir _build && cd _build
 cmake .. \
@@ -16,10 +28,10 @@ cmake .. \
     -DHDF5_NEED_SZIP=OFF \
     -DHDF5_NEED_ZLIB=ON \
     -DZLIB_LIBRARY=$BUILD_PREFIX/lib/libz$SHLIB_EXT \
-    -DHDF5_m_LIBRARY_RELEASE=$BUILD_PREFIX/$HOST/sysroot/usr/lib/libm$SHLIB_EXT \
-    -DHDF5_rt_LIBRARY_RELEASE=$BUILD_PREFIX/$HOST/sysroot/usr/lib/librt$SHLIB_EXT \
-    -DHDF5_dl_LIBRARY_RELEASE=$BUILD_PREFIX/$HOST/sysroot/usr/lib/libdl$SHLIB_EXT \
-    -DHDF5_pthread_LIBRARY_RELEASE=$BUILD_PREFIX/$HOST/sysroot/usr/lib/libpthread$SHLIB_EXT
+    -DHDF5_m_LIBRARY_RELEASE=$_libpath/libm$SHLIB_EXT \
+    -DHDF5_rt_LIBRARY_RELEASE=$_libpath/librt$SHLIB_EXT \
+    -DHDF5_dl_LIBRARY_RELEASE=$_libpath/libdl$SHLIB_EXT \
+    -DHDF5_pthread_LIBRARY_RELEASE=$_libpath/libpthread$SHLIB_EXT
 make -j$CPU_COUNT
 # ctest
 make install

--- a/recipe/macosx.patch
+++ b/recipe/macosx.patch
@@ -1,0 +1,13 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index 1c089f2..d4f1e62 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -357,6 +357,8 @@ set(CMAKE_INSTALL_RPATH "${CMAKE_INSTALL_PREFIX}/lib")
+ # which point to directories outside the build tree to the install RPATH
+ set(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
+ 
++set(CMAKE_MACOSX_RPATH TRUE)
++
+ #-----------------------------------------------------------------------------
+ # Dashboard and Testing Settings
+ #-----------------------------------------------------------------------------

--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -13,7 +13,7 @@ source:
     - macosx.patch
 
 build:
-  number: 0
+  number: 1
 
 requirements:
   build:

--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -10,6 +10,7 @@ source:
   git_rev: v{{ version }}
   patches:
     - interface.patch
+    - macosx.patch
 
 build:
   number: 0

